src/NumberTheory/vinberg.jl

```diff
@@ -353,3 +353,176 @@ function vinberg_algorithm(S::ZZLat, upper_bound; v0=QQ[0;]::QQMatrix, root_leng
   return [r * B for r in roots]
 end
 
+
+function _get_tau(f::ZZMatrix, Qb)
+  tau = zero(Qb)
+  for lambda in eigenvalues(Qb, f)
+    if abs(lambda)>tau
+      tau = abs(lambda)
+    end
+  end
+  return tau
+end
+
+function _get_bilinearform(Lf::ZZLatWithIsom, Qb)
+  gram_in_Qb = change_base_ring(Qb, gram_matrix(Lf))
+  #return (a, b)->inner_product(ambient_space(L), a,b)[1,1] -> it doesn't go well as there is no convienient way to convert ZZLat in QQ field to the Qb field
+  return function (a,b)
+    tmp_a = change_base_ring(Qb, a)
+    tmp_b = transpose(change_base_ring(Qb, b))
+    mul!(tmp_b, mul!(tmp_a, tmp_a, gram_in_Qb), tmp_b)
+    return tmp_b[1]
+  end
+end
+
+function _get_C0(Lf::ZZLatWithIsom, tau::QQBarFieldElem)
+  charPolyF = characteristic_polynomial(Lf)
+  x = gen(parent(charPolyF))
+  (n, remainder) = remove(charPolyF, x-1) # remove all x-1 factors
+  return div(remainder, minpoly(parent(charPolyF), tau)) #remove Salem polynomial of salem number tau
+end
+
+function _get_Cfancy(Lf::ZZLatWithIsom, C0)
+  return short_vectors_iterator(lattice(kernel_lattice(Lf, C0)), 2 , 2)
+end
+
+
+@doc raw"""
+    _get_h(L::ZZLat, v, w, Qb, bi_form) -> ZZMatrix
+
+Return a vector $h$ in the lattice `L`, such that $bi_form(h,h)>0$. 
+
+Vectors 'v','w' are eigenvectors of isometry associated with L based on salem number $tau$ and $tau^(-1)$
+
+'Qb' is a current field, where all calculations happen
+"""
+function _get_h(L::ZZLat, v, w, bi_form)
+  l = number_of_rows(basis_matrix(L))
+  r = rank(L)
+  if r<4
+    k = 10
+  elseif  r<6
+    k = 6
+  elseif r<8
+    k = 4
+  elseif r<10
+    k = 2
+  else
+    k = 1
+  end
+  h0 = v+w
+  h0 = h0/bi_form(h0,h0)
+  n = 1
+  i = 1
+  z = matrix(ZZ,1,l,rand(-k:k, l)) # random small vector in lattice basis
+  n_h0 = map(x->floor(ZZRingElem, x) , h0*n)
+  h = z+n_h0# in lattice basis
+  h = map(x->floor(ZZRingElem, x) , h) # in lattice basis and rounded
+  while bi_form(h,h) <= 0 || n == 10000  # block on 10000, because for a bigger n time consumption is way too big
+    i+=1
+    if i == 100
+      n+=1
+      i = 1
+      n_h0 = map(x->floor(ZZRingElem, x) , h0*n) 
+    end
+    for x in 1:l
+      z[1, x] = rand(-k:k)
+    end
+    add!(h, z, n_h0)
+    #h = (z+n_h0) #in lattice basis
+  end
+  if n == 10000 throw(OverflowError("The upper limitation on h calculation is reached. Bigger h will not be able to process in reasonable time")) end
+  return h::ZZMatrix
+end
+
+function _get_R(L, h::ZZMatrix)
+  return short_vectors_affine(change_base_ring(ZZ, gram_matrix(L)),h,0,-2) #better to use iterator, but there is no equivalent function that returns iterator
+end
+
+@doc raw"""
+    _process_finite_sets_of_h(h, f::QQMatrix, v, w, bi_form, L::ZZLat) -> Tuple{Bool, ZZMatrix}
+
+The function calculates all $r$, that can be obstructing roots of 'L'.
```

Comment:
```suggestion
The function calculates all $r$, that can be obstructing roots of `L`.
```
Changed to backticks.